### PR TITLE
Resolve duplicate intakes #186641727

### PR DESCRIPTION
**Pull request by tofarr**
* Allow the magic verification code 000000 for logins (We now allow it only when creating accounts)
* Fix the issue with locked accounts
* When you enter contact info that matches an existing intake, you validate your SSN and are taken into that intake

### CODE REVIEW

File: app/models/state_file_base_intake.rb [hunk 1]
```diff
@@ -198,7 +198,13 @@ def return_status
   def increment_failed_attempts
     super
     if attempts_exceeded?
-      lock_access! unless access_locked?
+      lock_access! if locking_enabled? and !access_locked?
```
Comment (tofarr) [inline app/models/state_file_base_intake.rb:201]: This patches the issue with accounts being locked.

File: app/models/state_file_base_intake.rb [hunk 2]
```diff
@@ -12,7 +12,7 @@ class StateFileBaseIntake < ApplicationRecord
   accepts_nested_attributes_for :primary_state_id, :spouse_state_id
 
   scope :can_be_authenticated, -> { where.not(hashed_ssn: nil) }
-  devise :timeoutable, :timeout_in => 15.minutes
+  devise :timeoutable, :timeout_in => 15.minutes, :unlock_strategy => :time
```
Comment (tofarr) [inline app/models/state_file_base_intake.rb:15]: This fixes the account locking issues - the unlock_strategy was `:none`

File: app/controllers/portal/client_logins_controller.rb [hunk 3]
```diff
@@ -35,9 +35,13 @@ def check_verification_code
         return
       end
 
-      @verification_code_form = Portal::VerificationCodeForm.new(contact_info: params[:contact_info], verification_code: params[:verification_code])
+      verification_code = params[:verification_code]
+      @verification_code_form = Portal::VerificationCodeForm.new(contact_info: params[:contact_info], verification_code: verification_code)
       if @verification_code_form.valid?
-        hashed_verification_code = VerificationCodeService.hash_verification_code_with_contact_info(params[:contact_info], params[:verification_code])
+        hashed_verification_code = VerificationCodeService.hash_verification_code_with_contact_info(params[:contact_info], verification_code)
+        if Rails.configuration.allow_magic_verification_code && @verification_code_form.verification_code == "000000"
```
[the remark this answers is not in the document]
Comment (embarnard) [reply app/controllers/portal/client_logins_controller.rb:42]: I checked if this works in the GYR portal login and it does! yay

File: app/controllers/state_file/questions/verification_code_controller.rb [hunk 4]
```diff
@@ -24,6 +25,53 @@ def edit
...
+          track_question_answer
+          redirect_to(next_path)
+        else
+          after_update_failure
+          track_validation_error
+          render :edit
+        end
+      end
+
+      def get_existing_intake(intake)
```
Comment (embarnard) [inline app/controllers/state_file/questions/verification_code_controller.rb:48]: should this be a private methods?

File: app/controllers/state_file/questions/verification_code_controller.rb [hunk 5]
```diff
@@ -24,6 +25,53 @@ def edit
...
+        search = intake.class.where.not(id: intake.id, raw_direct_file_data: nil)
+        search = search.where(phone_number: intake.phone_number) if intake.phone_number.present?
+        search = search.where(email_address: intake.email_address) if intake.email_address.present?
+        search.first
+      end
+
+      def redirect_into_login(intake, existing_intake)
```
Comment (embarnard) [inline app/controllers/state_file/questions/verification_code_controller.rb:55]: this one too

**Review by embarnard — APPROVED** [review]
Looks good! left some comments

File: app/controllers/portal/client_logins_controller.rb [hunk 6]
```diff
@@ -131,5 +135,22 @@ def gyr_redirect_unless_open_for_logged_in_clients
 
       redirect_to portal_closed_login_path unless open_for_gyr_logged_in_clients?
     end
+
+    def update_existing_token_with_magic_code(hashed_verification_code)
+      # If the environment supports magic codes, then the easiest thing is to
+      # update the last record with the magic code.
```
[the remark this answers is not in the document]
Comment (tofarr) [reply app/controllers/portal/client_logins_controller.rb:141]: Agreed!

File: app/controllers/state_file/questions/verification_code_controller.rb (hunk 4 above)
Comment (tofarr) [reply app/controllers/state_file/questions/verification_code_controller.rb:50]: Good call

**Pull request merged**